packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js

```diff
@@ -2428,17 +2440,21 @@ describe('ReactSuspenseWithNoopRenderer', () => {
       'Loading B...',
     ]);
     // Still suspended.
-    expect(ReactNoop.getChildren()).toEqual([span('A')]);
```

Comment:
The following are sources of the OOM problem. Interestingly, using 'yarn test' does not cause an OOM problem, but using the following commands does: 'yarn test -r=experimental --env=development --persistent --ci', 'yarn test -r=experimental --env=production --ci', 'yarn test --build -r=experimental --env=development --ci', 'yarn test --build -r=experimental --env=production --ci'. It is speculated that this may be because the OOM problem is avoided due to mocking of certain modules in the 'scripts/jest/setupTests.[www.js](http://www.js/)' file that is referenced in 'yarn test'.

Reply:
Why are we only using `removeNonEnumerableProperties` on some assertions? We have >100 assertions on `ReactNoop.getChildren()` but only a couple of are now wrapped in `removeNonEnumerableProperties`.

Reply:
Thanks for pointing out the issue with the assertions. 

I did some further investigation and found that the OOM problem was caused by the tests with the titles "keeps showing an avoided parent fallback if it is already showing" and "do not show placeholder when mounting an avoided boundary with startTransition".

I located the specific lines of code that were causing the problem and initially wrapped them with removeNonEnumerableProperties, but upon further review, I realized that this may not be the best solution. 

I think it would be better to further investigate and find the root cause of the problem.

Reply:
Did you get OOM locally or in CI?

Reply:
You can reproduce it locally with the ci test command

https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621083
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621073
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621072
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621071
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621053

Reply:
Not getting an OOM locally but without `removeEnumerableProperties` it takes 3 minutes for this test (to fail as expected).

Checking out what is wrong. Might be worth to convert all `expect(ReactNoop.getChildren()).toEqual` to `expect(ReactNoop).toMatchRenderedOutput` first. But we also need to understand what perf bottleneck we're hitting here.

Reply:
Seems like some perf regression in `expect` when `toEqual` does not pass. 

The "received" object contains fiber nodes which are nodes in a double linked list. And it seems like `jest-matcher-utils` is calling `deepCyclicCopyObject` (from `printDiffOrStringify`) in `node_modules/jest-matcher-utils/build/deepCyclicCopyReplaceable.js` thousands of times.

I'll create branch to highlight this for Jest people. 

Failing tests are intentionally part of our test suite. But if Jest 29 introduces such severe perf regressions (going from 50ms tests to 150000ms tests) for failing `toEqual` assertions, we can't merge it in this form.

Thank you very much for highlighting this issue!

Reply:
@SimenB 
```
$ git clone https://github.com/eps1lon/react.git
$ git checkout jestbump-perf-repro
$ yarn
$ yarn test -r=experimental --env=development --persistent --ci packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js --verbose
```

I'm also getting a lot of deprecation warnings from `testcheck` which is used by Jasmine. So maybe it's actually a Jasmine issue?

Reply:
I couldn't agree more with your concern about the performance degradation of the toEqual assertion, and will consider merging this pr after jest fixes the issue, very interested in following up on the jest performance issue fix process

Reply:
Before is also the best efforts to analyze the results of almost, but afraid to say not clear did not put forward. https://github.com/facebook/react/pull/26041#issuecomment-1406904681

Reply:
We probably have this problem on `main` (i.e. Jest 26) already: https://app.circleci.com/pipelines/github/facebook/react/38061/workflows/a2043dd0-df62-49d5-a10e-9b534c33cbdb/jobs/623788/parallel-runs/5?filterBy=ALL

Going to convert `expect(ReactNoop.getChildren()).toEqual` to `expect(ReactNoop).toMatchRenderedOutput` in a separate PR. This will hopefully fix issues on `main` and unblock this PR.

Reply:
Settled on using `toMatchRenderedOutput` instead.

Reply:
Is there a smaller reproduction than react?

Reply:
I don't have one and I expect it to be quite hard to reproduce. From what I could tell, the doubly linked-list structure of Fiber caused the severe slowdown in `expect(somethingHavingFibers).toEqual()`.